nexus/reconfigurator/planning/src/blueprint_builder.rs

```diff
@@ -275,49 +278,67 @@ impl<'a> BlueprintBuilder<'a> {
         // Note that by building these iterators up front based on
         // `parent_blueprint`, we cannot reuse resources in a case where we
         // remove a zone that used a resource and then add another zone that
-        // wants the same kind of resource. We don't support zone removal yet,
-        // but expect this to be okay: we don't anticipate removal and addition
-        // to frequently be combined into the exact same blueprint, particularly
-        // in a way that expects the addition to reuse resources from the
-        // removal; we won't want to attempt to reuse resources from a zone
-        // until we know it's been fully removed.
+        // wants the same kind of resource. That is mostly fine, but see
+        // the comment immediately below this block.
+
         let mut existing_nexus_v4_ips: HashSet<Ipv4Addr> = HashSet::new();
         let mut existing_nexus_v6_ips: HashSet<Ipv6Addr> = HashSet::new();
         let mut used_external_ips: HashSet<IpAddr> = HashSet::new();
         let mut used_macs: HashSet<MacAddr> = HashSet::new();
 
+        // XXX: the below comment is wrong, be more careful here. Need separate
+        // filters for internal and external IPs! Also need a test for this.
+        //
+        // Note: we're iterating over *all* omicron zones here, including
+        // expunged ones. For now, we don't unassign a resource from an
+        // expunged sled/zone and reassign it to a different zone in the same
+        // blueprint. (This can happen across successive blueprints, though.)
+        //
+        // Note, however, that we may want to reassign external IPs for
+        // external DNS zones within a single blueprint, but we're not solving
+        // that problem at the moment.
+
         for (_, z) in
-            parent_blueprint.all_omicron_zones(BlueprintZoneFilter::All)
+            parent_blueprint.all_blueprint_zones(BlueprintZoneFilter::All)
         {
-            let zone_type = &z.zone_type;
-            if let OmicronZoneType::Nexus { nic, .. } = zone_type {
-                match nic.ip {
-                    IpAddr::V4(ip) => {
-                        if !existing_nexus_v4_ips.insert(ip) {
-                            bail!("duplicate Nexus NIC IP: {ip}");
+            let zone_type = &z.config.zone_type;
+
+            // Running services have internal IPs.
+            if z.disposition.matches(BlueprintZoneFilter::ShouldBeRunning) {
```

Comment:
Is this correct? I'm a little worried that this isn't quite what we want, because it's possible that the last blueprint hasn't yet finished execution, and that the resources haven't actually been freed up, by the time we're generating the new one. (In that case, should resources allocated to expunged zones just live forever until the zones are garbage collected -- a process we aren't doing for now?)

Reply:
Yeah.... having thought more about it I'm pretty sure this is wrong, and that we actually need to get to the point where we've removed all records.

Reply:
Discussed this in a meeting yesterday -- we need to check the planning input to see if the zones are still there.

Reply:
In today's meeting we decided to just punting on garbage collection for release 8, which simplifies matters considerably.

Reply:
https://github.com/oxidecomputer/omicron/issues/5552 tracks this.